crates/eips/src/eip2718.rs

```diff
@@ -81,7 +83,7 @@ pub trait Decodable2718: Sized {
         let buf = &mut &buf[1..];
         let tx = Self::typed_decode(ty, buf)?;
 
-        if buf.len() != pre_len - h.payload_length {
+        if !buf.is_empty() {
```

Comment:
as far as I understand, this error is meant for when the RLP string has a payload length `n`, but the actual encoding ends up taking some `length != n`, so I'm not sure that this is the right way to fix. It might be worth checking what the `payload_length` actually is in the failing test.

Reply:
Interestingly, i checked the failing test (eip1559) and the payload length is 1—this seems incorrect, no?

Reply:
> Interestingly, i checked the failing test (eip1559) and the payload length is 1—this seems incorrect, no?

The `Header::decode` method is decoding the tx type byte then. IIRC if you try to `Header::decode` a single byte (less than `0x7f`), then `Header::decode` will:
* _NOT_ advance the buffer, and
* return successfully with `Header { payload_length: 1, list: false }`

Reply:
Not saying that `Header::decode` _should_ do this for the record, maybe `Header` should be an enum, or trying to decode a single byte should fail, but this is out of scope w.r.t. bug fixing:
```rust
enum Header {
    Byte{
        value: u8
    },
    ListHeader {
        payload_length: u64,
    },
    StringHeader {
        payload_length: u64,
    },
}
```

Reply:
Ah, that's right! this checks out with how we used to decode typed txs in anvil. I think, in this case, fallibly decoding the tx should be enough and this length check should not be required at all?

Reply:
This is how we do it in reth at least (see comments about quirks):
https://github.com/paradigmxyz/reth/blob/fd683e5a5c38e27aee2b7256f5cd270d17954019/crates/primitives/src/transaction/mod.rs#L1373-L1400

which also contains another check in `decode_enveloped_typed_transaction` for the inner typed tx rlp header payload length

this check should be performed IMO

Reply:
Awesome, sounds good. Corrected in https://github.com/alloy-rs/alloy/pull/148/commits/055a21192931348fc4ae6d658396089a2bff55bf